src/d3a/models/myco_matcher/external_matcher.py

```diff
@@ -0,0 +1,137 @@
+
+    def match_recommendations(self, message):
+        """
+        Receive trade recommendations and match them in the relevant market
+        """
+        channel = f"{self.channel_prefix}response/matched_recommendations/"
+        response_dict = {"event": "match", "status": "success"}
+        data = json.loads(message.get("data"))
+        recommendations = data.get("recommended_matches", [])
+        validated_records = {}
+        for record in recommendations:
+            market = self.markets_mapping.get(record.get("market_id"), None)
+            if market is None or market.readonly:
+                # The market is already finished or doesn't exist
+                del record
+                continue
+
+            bid = bid_from_JSON_string(json.dumps(record.get("bid")))
+            offer = offer_from_JSON_string(json.dumps(record.get("offer")),
+                                           record.get("bid").get("time"))
+
+            try:
+                validate_authentic_bid_offer_pair(
+                    bid,
+                    offer,
+                    record.get("trade_rate"),
+                    record.get("selected_energy")
+                    )
+                assert offer.id in market.offers
+                assert bid.id in market.bids
+                if record.get("market_id") not in validated_records:
+                    validated_records[record.get("market_id")] = []
+
+                validated_records[record.get("market_id")].append(BidOfferMatch(
+                    bid,
+                    record.get("selected_energy"),
+                    offer,
+                    record.get("trade_rate")))
+            except AssertionError:
+                # If validation fails or offer/bid were consumed
+                response_dict["status"] = "fail"
+                response_dict["message"] = "Validation Error"
+                break
+        if response_dict["status"] == "success":
+            for market_id, records in validated_records.items():
+                market = self.markets_mapping.get(market_id)
+                if market.readonly:
+                    # The market has just finished
+                    continue
+                market.match_recommendation(records)
+        self.myco_ext_conn.publish_json(channel, response_dict)
+
+    def get_simulation_id(self, message):
+        """
+        Publish the simulation id to the Myco client
+        """
+        channel = "external-myco/get_simulation_id/response"
+        self.myco_ext_conn.publish_json(channel, {"simulation_id": self.simulation_id})
+
+    def publish_event_tick_myco(self):
+        """
+        Myco API
+        """
+        channel = f"external-myco/{d3a.constants.COLLABORATION_ID}/response/events/"
+        data = {"event": "tick"}
+        self.myco_ext_conn.publish_json(channel, data)
+
+    def publish_market_cycle_myco(self):
+        """
+        Myco API
+        """
```

Comment:
- Nit: I'm not convinced that this docstring says much about the method (especially since it's the same docstring repeated for multiple methods).
- Nit: PEP 257 (the one for docstrings) recommends using a single line when possible. So this would be `"""Myco API"""`. Anyway, this is very minor and I appreciate the fact that you added documentation 👍